Show expected in AssertUtil.assertThrows

ashawley opened the pull request:
Error should mention what was expected exception:

    wrong exception: expected Foo but was Bar

Current test error is simply re-throwing the failed exception:

    failed: Bar

som-snytt commented on src/testkit/scala/tools/testkit/AssertUtil.scala:
```diff
@@ -92,6 +92,9 @@ object AssertUtil {
       case e: T if checkMessage(e.getMessage) =>
+      case t: Throwable                       =>
+        val clazz = scala.reflect.classTag[T].runtimeClass
+        fail(s"wrong exception: expected ${clazz.getName} but was ${t.getClass.getName}")
```
This has a slightly better formulation at L112 or so in `assertThrown`.

I think `assertThrows` could be written in terms of `assertThrown`.

som-snytt replied:
That is, I like that it uses `NonFatal` and suppressed exception. The test is called `discriminating`, enough said.

ashawley replied:
So it does.  Let me see if it can.

som-snytt commented on test/junit/scala/tools/testkit/AssertThrowsTest.scala:
```diff
@@ -31,14 +31,15 @@ class AssertThrowsTest {
       } catch {
-        case bar: Bar => true
-        case e: Throwable => fail(s"expected Bar but got $e"); false
+        case b: Bar => fail(s"Bar shouldn't have been rethrown"); false
```
We need `-Wunused:interpolator`!

ashawley replied:
Good catch!

ashawley committed:
Show expected in AssertUtil.assertThrows

Error should mention what was expected exception:

    wrong exception: expected Foo but was Bar

Current test error is simply re-throwing the failed exception:

    failed: Bar

som-snytt commented on test/junit/scala/tools/testkit/AssertThrowsTest.scala:
```diff
@@ -71,7 +72,7 @@ class AssertThrowsTest {
       case ae: AssertionError =>
         assertEquals(1, ae.getSuppressed.length)
-        assertEquals("Exception not a scala.tools.testkit.AssertThrowsTest$Foo: scala.tools.testkit.AssertThrowsTest$Bar", ae.getMessage)
+        assertEquals("Wrong exception: expected scala.tools.testkit.AssertThrowsTest$Foo but was scala.tools.testkit.AssertThrowsTest$Bar", ae.getMessage)
```
"It's [the wrong trousers](https://www.imdb.com/title/tt0108598/), Gromit!"

som-snytt reviewed (approved):
Thanks, very satisfying. I wish they had been more creative in naming `assertThrown`, or maybe less creative.
